src/cargo/ops/cargo_rustc/context.rs

```diff
@@ -332,7 +334,9 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
                 }
             }
         }
-        assert!(!ret.is_empty());
+        if ret.is_empty() {
+            bail!("no valid targets found")
```

Comment:
Perhaps this could be a more tailored error message? For example this could detect the case that a dylib output was requested but the target doesn't support it (and say that as part of the error)